08_irq/irq/gpio_irq.c

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/gpio.h>
+
+//#include <linux/gpio/driver.h>
+
+#include <linux/delay.h>
+#include <linux/moduleparam.h>
+#include <linux/interrupt.h>
+#include <linux/irqflags.h>
+
+MODULE_AUTHOR("Andrii Synenko");
+MODULE_DESCRIPTION("List data module for Linux Kernel ProCamp");
+MODULE_LICENSE("Dual BSD/GPL");
+
+#define GPIO_NUMBER(port, bit) (32 * (port) + (bit))
+
+static u32 state_led_pin;
+module_param(state_led_pin, uint, 0);
+MODULE_PARM_DESC(state_led_pin, "LED pin ID");
+
+static u32 user_btn_pin;
+module_param(user_btn_pin, uint, 0);
+MODULE_PARM_DESC(user_btn_pin, "User button pin ID");
+
+static u32 busy_led_pin;
+module_param(busy_led_pin, uint, 0);
+MODULE_PARM_DESC(busy_led_pin, "Busy LED pin ID");
+
+static u32 simulate_busy;
+module_param(simulate_busy, uint, 0);
+MODULE_PARM_DESC(simulate_busy, "Sinulation delay in ms");
+
+struct gpio_conf {
+	u32 pin_id;
+	bool output;
+
+	int irq_id;
+	irq_handler_t irq_handler;
+	irq_handler_t irq_thread;
+};
+
+static struct gpio_conf user_btn  = {66, false, -1, NULL, NULL};
+static struct gpio_conf state_led = {67, true, -1, NULL, NULL};
+static struct gpio_conf busy_led  = {68, true, -1, NULL, NULL};
+static u32 hit_counter;
+
+static int gpio_init(struct gpio_conf *pin)
+{
+	int res = 0;
+
+	res = gpio_request(pin->pin_id, "Onboard user button");
+	if (res < 0) {
+		pr_err("%s: failed to request gpio pin: %d",
+		       __func__, pin->pin_id);
+		return res;
+	}
+
+	res = (pin->output == true) ? gpio_direction_output(pin->pin_id, 0)
+				: gpio_direction_input(pin->pin_id);
+	if (res < 0)
+		goto failed_gpio_init;
+
+	if (pin->output == false)
+		gpio_set_debounce(pin->pin_id, 20);
```

Comment:
20 microseconds is too short time. Perhaps, 2000 is OK in most cases.

Reply:
Fixed